File: verify_project.py

import os
import re
import json

def strip_comments(text):
    return re.sub(r'//.*', '', text)

def get_keys_from_js(content, lang):
    # This is a heuristic parser. 
    # It assumes the file structure:
    # en: { ... }, ua: { ... }
    
    # Find the start of the language block
    start_pattern = f'{lang}:' + r'\s*\{'
    start_match = re.search(start_pattern, content)
    if not start_match:
        return set()
    
    start_index = start_match.end()
    
    # Find the matching closing brace (simple counter)
    brace_count = 1
    end_index = start_index
    for i, char in enumerate(content[start_index:]):
        if char == '{':
            brace_count += 1
        elif char == '}':
            brace_count -= 1
        
        if brace_count == 0:
            end_index = start_index + i
            break
            
    block_content = content[start_index:end_index]
    
    # Extract keys
    # Keys are quoted strings followed by a colon
    keys = set()
    # We strip comments first
    clean_block = re.sub(r'//.*', '', block_content)
    
    for match in re.finditer(r'"([^"]+)"\s*:', clean_block):
        keys.add(match.group(1))
        
    return keys

def main():
    # 1. Load Translations
    with open('translations.js', 'r', encoding='utf-8') as f:
        content = f.read()
        
    en_keys = get_keys_from_js(content, 'en')
    ua_keys = get_keys_from_js(content, 'ua')
    
    print(f"Loaded {len(en_keys)} EN keys and {len(ua_keys)} UA keys.")
    
    # Check consistency
    missing_in_ua = en_keys - ua_keys
    missing_in_en = ua_keys - en_keys
    
    if missing_in_ua:
        print("Keys missing in UA:", missing_in_ua)
    if missing_in_en:
        print("Keys missing in EN:", missing_in_en)

    # 2. Scan HTML files
    used_keys = set()
    broken_links = []
    
    root_dir = os.getcwd()
    
    for dirpath, dirnames, filenames in os.walk(root_dir):
        # Skip node_modules and .git
        if 'node_modules' in dirnames:
            dirnames.remove('node_modules')
        if '.git' in dirnames:
            dirnames.remove('.git')
            
        for filename in filenames:
            if not filename.endswith('.html'):
                continue
                
            filepath = os.path.join(dirpath, filename)
            rel_path = os.path.relpath(filepath, root_dir)
            
            with open(filepath, 'r', encoding='utf-8') as f:
                file_content = f.read()
                
            # Find data-i18n attributes
            # data-i18n="key", data-i18n-tooltip="key", etc.
            # We look for any attribute starting with data-i18n
            for match in re.finditer(r'data-i18n(?:-[a-z]+)?="([^"]+)"', file_content):
                used_keys.add(match.group(1))
                
            # Check links
            for match in re.finditer(r'href="([^"]+)"', file_content):
                link = match.group(1)
                if link.startswith(('#', 'http', 'mailto:', 'javascript:')):
                    continue
                
                # Resolve link
                # Remove anchors and query params
                clean_link = link.split('#')[0].split('?')[0]
                if not clean_link:
                    continue
                    
                target_path = os.path.normpath(os.path.join(dirpath, clean_link))
                
                if not os.path.exists(target_path):
                     broken_links.append((rel_path, link))

    # 3. Report
    missing_keys = used_keys - en_keys
    if missing_keys:
        print("\nMissing translation keys (used in HTML but not in translations.js):")
        for k in missing_keys:
            print(f" - {k}")
    else:
        print("\nAll used translation keys are present.")
        
    if broken_links:
        print("\nBroken internal links:")
        for file, link in broken_links:
            print(f" - In {file}: {link}")
    else:
        print("\nNo broken internal links found.")

if __name__ == "__main__":
    main()

File: verify_project.js

const fs = require('fs');
const path = require('path');

// 1. Load Translations
const translationsContent = fs.readFileSync('translations.js', 'utf8');
const getTranslations = new Function(translationsContent + '; return translations;');
const translations = getTranslations();

const languages = Object.keys(translations);
console.log(`Loaded languages: ${languages.join(', ')}`);

const enKeys = new Set(Object.keys(translations['en']));
const uaKeys = new Set(Object.keys(translations['ua']));

// Check for consistency between languages
const missingInUa = [...enKeys].filter(k => !uaKeys.has(k));
const missingInEn = [...uaKeys].filter(k => !enKeys.has(k));

if (missingInUa.length > 0) console.log('Keys missing in UA:', missingInUa);
if (missingInEn.length > 0) console.log('Keys missing in EN:', missingInEn);

// 2. Scan HTML files recursively
function getAllFiles(dirPath, arrayOfFiles) {
  files = fs.readdirSync(dirPath);

  arrayOfFiles = arrayOfFiles || [];

  files.forEach(function(file) {
    if (fs.statSync(dirPath + "/" + file).isDirectory()) {
      if (file !== 'node_modules' && file !== '.git') {
        arrayOfFiles = getAllFiles(dirPath + "/" + file, arrayOfFiles);
      }
    } else {
      if (file.endsWith('.html')) {
        arrayOfFiles.push(path.join(dirPath, "/", file));
      }
    }
  });

  return arrayOfFiles;
}

const htmlFiles = getAllFiles(__dirname);
const usedKeys = new Set();
const brokenLinks = [];

htmlFiles.forEach(file => {
  const content = fs.readFileSync(file, 'utf8');
  const dir = path.dirname(file);

  // Regex for data-i18n attributes
  const regexes = [
    /data-i18n="([^"]+)"/g,
    /data-i18n-tooltip="([^"]+)"/g,
    /data-i18n-placeholder="([^"]+)"/g,
    /data-i18n-title="([^"]+)"/g
  ];

  regexes.forEach(regex => {
    let match;
    while ((match = regex.exec(content)) !== null) {
      usedKeys.add(match[1]);
    }
  });

  // Check links
  const linkRegex = /href="([^"]+)"/g;
  let linkMatch;
  while ((linkMatch = linkRegex.exec(content)) !== null) {
    const link = linkMatch[1];
    if (link.startsWith('#') || link.startsWith('http') || link.startsWith('mailto:') || link.startsWith('javascript:')) continue;
    
    // Construct absolute path
    let targetPath = path.resolve(dir, link);
    
    // Remove query params or anchors
    targetPath = targetPath.split('#')[0].split('?')[0];

    if (!fs.existsSync(targetPath)) {
         brokenLinks.push({ file: path.relative(__dirname, file), link });
    }
  }
});

// 3. Report Missing Keys
const missingKeys = [...usedKeys].filter(k => !enKeys.has(k));
if (missingKeys.length > 0) {
  console.log('Missing translation keys (used in HTML but not in translations.js):');
  missingKeys.forEach(k => console.log(` - ${k}`));
} else {
  console.log('All used translation keys are present.');
}

if (brokenLinks.length > 0) {
    console.log('Broken internal links:');
    brokenLinks.forEach(item => console.log(` - In ${item.file}: ${item.link}`));
} else {
    console.log('No broken internal links found.');
}
